### PR TITLE
adding the possibility to test all defined dao-queries

### PR DESCRIPTION
**Pull request by rassmate**
Added a test that searches the classpath for dao classes and executes the dao queries against a database initialized with all the migration files it can find. 

Creating random not null objects and pass them to the dao layer to ensure the sql syntax is valid.

To use it in your module just make a subclass of the DaoTester in your module and the testing will be carried out. There is also a possibility to customize the randomized objects that should be used by overriding methods in the DaoTester

### CODE REVIEW

File: dbmigrate/src/test/java/se/fortnox/reactivewizard/dbmigrate/DaoTester.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Enumeration;
+
+import static se.fortnox.reactivewizard.test.TypeRandomizer.getType;
+
+/**
+ * Sublass this class and create a single test calling the method DaoTester#testDaoClasses
```
Comment (flowertwig) [inline dbmigrate/src/test/java/se/fortnox/reactivewizard/dbmigrate/DaoTester.java]: Subclass

File: dbmigrate/src/test/java/se/fortnox/reactivewizard/dbmigrate/DaoTester.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+            mockValues.add(getRandomValue(daoClass, method, index, parameter.getType()));
+            index++;
+        }
+
+        AssertableSubscriber<?> test = ((Observable<?>)method.invoke(daoInstance, mockValues.toArray())).onErrorResumeNext(throwable -> {
+            throwable.printStackTrace();
+            return Observable.error(throwable);
+        }).test();
+        test.awaitTerminalEvent();
```
Comment (flowertwig) [inline dbmigrate/src/test/java/se/fortnox/reactivewizard/dbmigrate/DaoTester.java]: Inline?

File: bootstrap/pom.xml [hunk 3]
```diff
@@ -15,8 +15,20 @@
 			<groupId>com.google.inject</groupId>
 			<artifactId>guice</artifactId>
 			<version>${guice.version}</version>
+            <exclusions>
```
Comment (jepp3) [inline bootstrap/pom.xml]: is the exlusion really needed if the version is specified in the pom file as well?

File: dbmigrate/pom.xml [hunk 4]
```diff
@@ -80,6 +87,40 @@
             <scope>test</scope>
             <version>1.0.0-SNAPSHOT</version>
         </dependency>
+
+        <dependency>
+            <groupId>org.testcontainers</groupId>
```
Comment (jepp3) [inline dbmigrate/pom.xml]: we will depend on docker for executing tests now. :)

File: dbmigrate/src/test/java/se/fortnox/reactivewizard/dbmigrate/DaoTester.java (hunk 1 above)
Comment (jepp3) [reply dbmigrate/src/test/java/se/fortnox/reactivewizard/dbmigrate/DaoTester.java]: Do you mean extend?

File: test/src/test/java/se/fortnox/reactivewizard/test/TypeRandomizerTest.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
+package se.fortnox.reactivewizard.test;
+
+import org.junit.Test;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static org.fest.assertions.Assertions.assertThat;
+
+public class TypeRandomizerTest {
```
Comment (jepp3) [inline test/src/test/java/se/fortnox/reactivewizard/test/TypeRandomizerTest.java]: naming of test cases

File: test/src/test/java/se/fortnox/reactivewizard/test/TypeRandomizerTest.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    @Test(expected = RuntimeException.class)
+    public void testTypeRandomizerThrowsExceptionWhenClassCanNotBeInstantiated() {
+        TypeRandomizer.getType(AbstractPojo.class);
+    }
+
+    @Test
+    public void testTypeRandomizer() throws InvocationTargetException, IllegalAccessException {
```
Comment (jepp3) [inline test/src/test/java/se/fortnox/reactivewizard/test/TypeRandomizerTest.java]: what does it test

File: test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Supplier;
+
+public class TypeRandomizer {
+
+    private static final Random random = new Random();
+    private static Map<Class<?>, Supplier<?>>  suppliers = ImmutableMap.<Class<?>, Supplier<?>>builder()
```
Comment (jepp3) [inline test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java]: final

File: test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        }
+
+        if (suppliers.containsKey(type)) {
+            return (T)suppliers.get(type).get();
+        }
+
+        return createAndFill(type);
+    }
+
+    private static <T> T createAndFill(Class<T> clazz) {
```
Comment (jepp3) [inline test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java]: maybe a better name for the method. That describes more what it does

File: test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            }
+
+            return instance;
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalArgumentException("Could not instantiate class: " + clazz + " " + e.getMessage());
+        }
+    }
+
+    private static <T> void handleField(T instance, Field field) throws IllegalAccessException {
+        if (field.getName().startsWith("$")) {
```
Comment (jepp3) [inline test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java]: what is $?

File: test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+                }
+            }
+
+            return instance;
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalArgumentException("Could not instantiate class: " + clazz + " " + e.getMessage());
+        }
+    }
+
+    private static <T> void handleField(T instance, Field field) throws IllegalAccessException {
```
Comment (jepp3) [inline test/src/main/java/se/fortnox/reactivewizard/test/TypeRandomizer.java]: a name that states what the method does

**Comment by jepp3** [comment]
I think there are libraries that can do the generation of data if we dont want to do that our self. j-easy/easy-random for example.

**Comment by rassmate** [comment]
Wont be doing more work on this

**Pull request closed without merging**